File: src/cli/project/issues.ts
import type { Command } from "commander";
import type { LinearDocument } from "@linear/sdk";
import { getClient } from "../../lib/client.ts";
import { buildIssueFilter } from "../../lib/filters.ts";
import { printError, printPaginated, resolvePageSize } from "../../lib/output.ts";
import { resolveProject } from "../../lib/resolvers.ts";

export function registerIssues(project: Command): void {
  project
    .command("issues")
    .description("List issues within a project")
    .argument("<id>", "Project ID, slug, or name")
    .option("--status <status>", "Filter by status")
    .option("--assignee <user>", "Filter by assignee")
    .option("--priority <priority>", "Filter by priority")
    .option("--limit <n>", "Limit results", "50")
    .option("--cursor <token>", "Pagination cursor for next page")
    .action(
      async (
        id: string,
        opts: {
          status?: string;
          assignee?: string;
          priority?: string;
          limit: string;
          cursor?: string;
        },
      ) => {
        try {
          const client = getClient();
          const p = await resolveProject(client, id);
          const filter = buildIssueFilter(opts);
          const issues = await p.issues({
            first: resolvePageSize(opts),
            after: opts.cursor,
            filter:
              Object.keys(filter).length > 0 ? (filter as LinearDocument.IssueFilter) : undefined,
          });
          const items = await Promise.all(
            issues.nodes.map(async (i) => {
              const [state, assignee] = await Promise.all([i.state, i.assignee]);
              return {
                id: i.id,
                identifier: i.identifier,
                title: i.title,
                status: state ? state.name : null,
                assignee: assignee ? assignee.name : null,
                priority: i.priority,
                priorityLabel: i.priorityLabel,
              };
            }),
          );
          printPaginated(items, issues.pageInfo);
        } catch (err) {
          printError(err instanceof Error ? err.message : "Get issues failed");
        }
      },
    );
}

File: internal/cli/project/issues.go
package project

import (
	"context"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/filters"
	"github.com/shhac/lin/internal/linear"
	"github.com/shhac/lin/internal/mappers"
	"github.com/shhac/lin/internal/output"
	"github.com/shhac/lin/internal/resolvers"
)

func registerIssues(parent *cobra.Command) {
	var (
		status   string
		assignee string
		priority string
	)

	cmd := &cobra.Command{
		Use:   "issues <id>",
		Short: "List issues within a project",
		Args:  cobra.ExactArgs(1),
	}
	page := output.AddPageFlags(cmd)

	cmd.Run = func(cmd *cobra.Command, args []string) {
		client := linear.GetClient()
		ctx := context.Background()

		resolved, err := resolvers.ResolveProject(client, args[0])
		if err != nil {
			output.PrintError(err.Error())
		}

		filter := filters.BuildIssueFilter(filters.IssueFilterOpts{
			Assignee: assignee,
			Status:   status,
			Priority: priority,
		})

		resp, err := linear.ProjectIssues(ctx, client, resolved.ID, filter, page.Size(), page.Cursor())
		if err != nil {
			output.HandleGraphQLError(err)
		}

		items := make([]any, len(resp.Project.Issues.Nodes))
		for i, n := range resp.Project.Issues.Nodes {
			items[i] = mappers.MapIssueSummary(mappers.FromIssueSummaryFields(n.IssueSummaryFields))
		}

		output.PrintPage(items, resp.Project.Issues.PageInfo.HasNextPage, resp.Project.Issues.PageInfo.EndCursor)
	}

	cmd.Flags().StringVar(&status, "status", "", "Filter by status")
	cmd.Flags().StringVar(&assignee, "assignee", "", "Filter by assignee")
	cmd.Flags().StringVar(&priority, "priority", "", "Filter by priority")
	parent.AddCommand(cmd)
}
